### PR TITLE
feat(financial-product-library): Create a covered call financial product library

### PR DESCRIPTION
**Pull request by hlambur**
**Motivation**

To create covered call options, we require a simple library to transform the price of an identifier based on a strike price of the option. This PR adds the ability for an EMP to implement a covered call.

**Financial specification**
The contract holds say 1 WETH in collateral and pays out a portion of that, at expiry, if ETHUSD is above a set strike. If ETHUSD is below that strike, the contract pays out 0. The fraction paid out if above the strike is defined by `(oraclePrice - strikePrice) / oraclePrice`.

_Example_: expiry is DEC 31. Strike is $400. Each token is backed by 1 WETH.
- If ETHUSD = $600 at expiry, the call is $200 in the money, and the contract pays out 0.333 WETH (worth $200).
- If ETHUSD = $800 at expiry, the call is $400 in the money, and the contract pays out 0.5 WETH (worth $400).
- If ETHUSD =< $400 at expiry, the call is out of the money, and the contract pays out 0 WETH.

**Summary**

`CoveredCallFinancialProductLibrary.sol` added and associated unit tests.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [X]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/core/contracts/financial-templates/common/financial-product-libraries/CoveredCallFinancialProductLibrary.sol [hunk 1]
```diff
@@ -70,17 +71,16 @@ contract StructuredNoteFinancialProductLibrary is FinancialProductLibrary, Ownab
...
         if (oraclePrice.isLessThan(strike)) {
-            return FixedPoint.fromUnscaledUint(1);
+            return 0;
         } else {
-            // Token expires to be worth strike $ worth of collateral.
-            // eg if ETHUSD is $500 and strike is $400, token is redeemable for 400/500 = 0.8 WETH.
-            return strike.div(oraclePrice);
+            // Token expires to be worth the fraction of an UMA token that's in the money.
+            // eg if ETHUSD is $500 and strike is $400, token is redeemable for 100/500 = 0.2 WETH (worth $100).
+            return (orcalePrice - strike).div(oraclePrice);
```
Comment (mrice32) [inline packages/core/contracts/financial-templates/common/financial-product-libraries/CoveredCallFinancialProductLibrary.sol:78]:
```suggestion
            return (orcalePrice.sub(strike)).div(oraclePrice);
```

File: packages/core/contracts/financial-templates/common/financial-product-libraries/CoveredCallFinancialProductLibrary.sol [hunk 2]
```diff
@@ -91,17 +91,8 @@ contract StructuredNoteFinancialProductLibrary is FinancialProductLibrary, Ownab
...
-            // a strike of $400 and a CR of 1.2 would yield:
-            // ETHUSD = $350, payout is 1 WETH. CR is multiplied by 1. resulting CR = 1.2
-            // ETHUSD = $400, payout is 1 WETH. CR is multiplied by 1. resulting CR = 1.2
-            // ETHUSD = $425, payout is 0.941 WETH (worth $400). CR is multiplied by 0.941. resulting CR = 1.1292
-            // ETHUSD = $500, payout is 0.8 WETH (worth $400). CR multiplied by 0.8. resulting CR = 0.96
-            return collateralRequirement.mul(strike.div(oraclePrice));
-        }
+        
+        // Always return 1 because option must be collateralized by 1 token
+        return 1;
```
Comment (mrice32) [inline packages/core/contracts/financial-templates/common/financial-product-libraries/CoveredCallFinancialProductLibrary.sol:96]:
```suggestion
        return FixedPoint.fromUnscaledUint(1);
```

**Review by mrice32 — COMMENTED** [review]
This looks awesome! I think this is _pretty much_ what we want (I made a few small suggestions). Will review again tomorrow, but I think this is basically done.

One minor note: editing this contract is helpful to see the changes, but when you're ready to merge, make sure you put back the original contract -- otherwise the tests won't pass.

File: packages/core/contracts/financial-templates/common/financial-product-libraries/CoveredCallFinancialProductLibrary.sol [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+ * contract holds say 1 WETH in collateral and pays out a portion of that, at expiry, if ETHUSD is above a set strike. If
+ * ETHUSD is below that strike, the contract pays out 0. The fraction paid out if above the strike is defined by
+ * (oraclePrice - strikePrice) / oraclePrice;
+ * Example: expiry is DEC 31. Strike is $400. Each token is backed by 1 WETH.
+ * If ETHUSD = $600 at expiry, the call is $200 in the money, and the contract pays out 0.333 WETH (worth $200).
+ * If ETHUSD = $800 at expiry, the call is $400 in the money, and the contract pays out 0.5 WETH (worth $400).
+ * If ETHUSD =< $400 at expiry, the call is out of the money, and the contract pays out 0 WETH.
+ */
+contract CoveredCallFinancialProductLibrary is FinancialProductLibrary, Ownable, Lockable {
+    mapping(address => FixedPoint.Unsigned) financialProductStrikes;
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/common/financial-product-libraries/CoveredCallFinancialProductLibrary.sol:19]: Should this be explicitly a `private` mapping?

File: packages/core/contracts/financial-templates/common/financial-product-libraries/CoveredCallFinancialProductLibrary.sol [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        // each token backed 1:1 by collateral currency.
+        if (requestTime < ExpiringContractInterface(msg.sender).expirationTimestamp()) {
+            return FixedPoint.fromUnscaledUint(1);
+        }
+        if (oraclePrice.isLessThanOrEqual(strike)) {
+            return FixedPoint.fromUnscaledUint(0);
+        } else {
+            // Token expires to be worth the fraction of an collateral token that's in the money.
+            // eg if ETHUSD is $500 and strike is $400, token is redeemable for 100/500 = 0.2 WETH (worth $100).
+            return (oraclePrice.sub(strike)).div(oraclePrice);
```
Comment (nicholaspai) [inline packages/core/contracts/financial-templates/common/financial-product-libraries/CoveredCallFinancialProductLibrary.sol:79]: Is it worth adding a comment either here that we're assuming that `oraclePrice != 0`? [The EMP does allow the `oraclePrice` to be set to 0.](https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/financial-templates/expiring-multiparty/PricelessPositionManager.sol#L829)

**Review by nicholaspai — APPROVED** [review]
My comments can probably be addressed just by adding some code comments, looks awesome!

The `CoveredCall.sol` contract is almost identical to the `StructuredNote.sol` [contract](https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/financial-templates/common/financial-product-libraries/StructuredNoteFinancialProductLibrary.sol) FYI.

File: packages/core/contracts/financial-templates/common/financial-product-libraries/CoveredCallFinancialProductLibrary.sol (hunk 4 above)
Comment (mrice32) [reply packages/core/contracts/financial-templates/common/financial-product-libraries/CoveredCallFinancialProductLibrary.sol:79]: It's a great point. It definitely deserves a comment. I think it's impossible to get here if oraclePrice == 0.

If I'm thinking about this correctly, if oraclePrice == 0, then strike is always >= oraclePrice, and the if above is always true, meaning this function just returns 0. Do you agree?

**Review by mrice32 — APPROVED** [review]
LGTM outside of @nicholaspai's comments!

**Review by nicholaspai — APPROVED** [review]

**Pull request merged**